### PR TITLE
Add support of the new command SINTERCARD(Redis 7)

### LINKED ISSUE #1083: [NEW] Add `SINTERCARD` command aligned Redis
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-kvrocks/issues) and found no similar issues.

### Motivation

add support for [SINTERCARD](https://redis.io/commands/sintercard/) command in Redis, which is available since Redis 7.0.0

### Solution

Since `SINTER` and `SCARD` were implemented already, I think it's quite straightforward to combine those two commands' logic.

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by infdahai**
closes #1083
closes #1116 (replace)

This PR introduces a `sintercard` command to `kvrocks` from redis 7.0.

~BTW, the `go-redis` package's version is `v9.0.0-beta.2` in the project so it doesn't support this command(more infomation in https://github.com/redis/go-redis/pull/2291). And I just write a bare test in `go-test` now.~

~Perhaps we can bump `go-redis` to a newer version.~

### CODE REVIEW

File: src/types/redis_set.cc [hunk 1]
```diff
@@ -338,19 +338,42 @@ rocksdb::Status Set::Inter(const std::vector<Slice> &keys, std::vector<std::stri
...
     s = Members(keys[i], &target_members);
     if (!s.ok() || target_members.empty()) return s;
     for (const auto &member : target_members) {
-      if (member_counters.find(member) == member_counters.end()) continue;
+      if (member_counters.count(member) == 0) continue;
       member_counters[member]++;
+      if (has_limit && member_counters[member] == keys.size()) {
+        members->emplace_back(member);
+        if (--limit == 0) {
+          limit = -1;
```
Comment (torwig) [inline src/types/redis_set.cc]: The variable `limit` is of type `uint64_t`. How it could be assigned to `-1` and later checked via the condition `limit < 0`?

Comment (torwig) [reply src/types/redis_set.cc]: Maybe the following code should be considered as a loop break:

```
if (has_limit && member_counters[member] == keys.size()) {
      members->emplace_back(member);
      if (members->size() >= limit) {
        break;
      }
}
```

File: tests/cppunit/types/set_test.cc [hunk 2]
```diff
@@ -191,6 +191,23 @@ TEST_F(RedisSetTest, Inter) {
   set_->Del(k3);
 }
 
+TEST_F(RedisSetTest, InterCard) {
```
Comment (torwig) [inline tests/cppunit/types/set_test.cc:205]: Additionally, you can add a test for a corner/edge case when one of the sets is empty. And one for the case when there are no common elements in both sets.

File: src/types/redis_set.cc (hunk 1 above)
Comment (infdahai) [reply src/types/redis_set.cc]: oh, I forgot this type.

File: tests/cppunit/types/set_test.cc (hunk 2 above)
Comment (infdahai) [reply tests/cppunit/types/set_test.cc:205]: I plan to add these in `inter` tests later beacause `intercard` is a subset of calling `inter` .

File: src/commands/cmd_set.cc [hunk 3]
```diff
@@ -283,6 +285,46 @@ class CommandSInter : public Commander {
...
+  }
+
+  Status Execute(Server *svr, Connection *conn, std::string *output) override {
+    std::vector<Slice> keys;
+    for (size_t i = 2; i < numkeys_ + 2; i++) {
+      keys.emplace_back(args_[i]);
+    }
+
+    redis::Set set_db(svr->storage, conn->GetNamespace());
+    int ret = 0;
```
Comment (torwig) [inline src/commands/cmd_set.cc]: You don't use this `ret` variable, you don't send any response to the command. Please, add some integration tests in Go to check the proper responses of the newly created command.

Comment (infdahai) [reply src/commands/cmd_set.cc]: > You don't use this ret variable

yeah, I add it now.

>  the go-redis package's version is v9.0.0-beta.2 in the project so it doesn't support this command

So we should wait for updating the go-redis package and then write go tests.

Comment (torwig) [reply src/commands/cmd_set.cc]: You can just use `Do()` function of the client to send raw request instead of the command, for example: https://github.com/apache/incubator-kvrocks/blob/unstable/tests/gocase/unit/type/stream/stream_test.go#L45. Or update the package.

Comment (infdahai) [reply src/commands/cmd_set.cc]: OK, I get it.

**Comment by infdahai** [comment]
@git-hulk I haven't finished go tests because of waiting for https://github.com/apache/incubator-kvrocks/pull/1446. If I write it done, I will ask for reviewers.

File: src/types/redis_set.cc [hunk 4]
```diff
@@ -338,19 +338,43 @@ rocksdb::Status Set::Inter(const std::vector<Slice> &keys, std::vector<std::stri
...
+  bool has_limit = limit != 0;
+  bool limited = false;
   for (size_t i = 1; i < keys.size(); i++) {
     s = Members(keys[i], &target_members);
     if (!s.ok() || target_members.empty()) return s;
     for (const auto &member : target_members) {
-      if (member_counters.find(member) == member_counters.end()) continue;
+      if (member_counters.count(member) == 0) continue;
       member_counters[member]++;
+      if (has_limit && member_counters[member] == keys.size()) {
```
Comment (PragmaTwice) [inline src/types/redis_set.cc]: Why do we search for the `member` element three times in `member_counters` continuously? It looks so bad to me.

```c++
auto iter = find(member); // find it
use1(iter); // use it
use2(iter); // use it again

// **NOT**:
use1(find(member));
use2(find(member));
use3(find(member));
```

File: src/types/redis_set.cc [hunk 5]
```diff
@@ -338,19 +338,43 @@ rocksdb::Status Set::Inter(const std::vector<Slice> &keys, std::vector<std::stri
...
+      if (iter.second == keys.size()) {  // all the sets contain this member
+        members->emplace_back(iter.first);
+      }
     }
   }
+
+  return rocksdb::Status::OK();
+}
+
+rocksdb::Status Set::InterCard(const std::vector<Slice> &keys, uint64_t limit, int *ret) {
```
Comment (PragmaTwice) [inline src/types/redis_set.cc]: Maybe not related to this PR, but there are two points:
- `ret` is not a descriptive word
- `int` type is not fit here: the return variable `members.size()` is typed `size_t`, not `int`

File: tests/gocase/unit/type/set/set_test.go [hunk 6]
```diff
@@ -282,6 +284,8 @@ func TestSet(t *testing.T) {
 			}
 		})
 
+		t.Run("SINTERCARD against three sets - "+dstype, func(t *testing.T) {})
```
Comment (PragmaTwice) [inline tests/gocase/unit/type/set/set_test.go]: maybe we can add more tests here.

File: src/types/redis_set.cc (hunk 5 above)
Comment (infdahai) [reply src/types/redis_set.cc]: We find a few functions with `int *ret` in [redis_set.cc](https://github.com/apache/incubator-kvrocks/blob/unstable/src/types/redis_set.cc). However the result of `ret` is unsinged long(size_t). I will check the correct size type in multiple files.

Comment (infdahai) [reply src/types/redis_set.cc]: @PragmaTwice I exec the statement `*output = redis::Integer(ret);` to return the result to the client in cmd_set.cc.
https://github.com/apache/incubator-kvrocks/blob/fb0e3d42031a6e0691c7a736711d0a01d72bdb1a/src/server/redis_reply.h#L38-L41

Add the `Integer` func checks whether compatibility with `int` is required(Of course, `size_t` is right). I see src/types/*.h and many funcs with this situation use `int`.

File: src/commands/cmd_set.cc [hunk 7]
```diff
@@ -283,6 +285,55 @@ class CommandSInter : public Commander {
   }
 };
 
+/*
+ * description:
+ *
+ *    syntax: SINTERCARD numkeys key [key ...] [LIMIT limit]
+ *
+ *    params: the checked cardinality is unlimited (limit <= 0).
```
Comment (torwig) [inline src/commands/cmd_set.cc]: Redis prohibits the negative `LIMIT` parameter:

```
127.0.0.1:6379> sintercard 2 set1 set2 limit -5
(error) ERR LIMIT can't be negative
127.0.0.1:6379> 
```

BTW: right now the phrase `the checked cardinality is unlimited` sounds weird to me. 
Since `Kvrocks` supports `Redis` protocol, if you want to add some documentation, you can just copy-paste sentences directly from `Redis` docs.

Comment (infdahai) [reply src/commands/cmd_set.cc]: Thanks, I got it. I just read the docs about this command so I make a mistake.

**Comment by git-hulk** [comment]
The flaky test still exists after applying #1449 and it should be caused by the retry in Go redis client.

File: src/commands/cmd_set.cc [hunk 8]
```diff
@@ -371,6 +418,7 @@ REDIS_REGISTER_COMMANDS(MakeCmdAttr<CommandSAdd>("sadd", -3, "write", 1, 1, 1),
                         MakeCmdAttr<CommandSDiff>("sdiff", -2, "read-only", 1, -1, 1),
                         MakeCmdAttr<CommandSUnion>("sunion", -2, "read-only", 1, -1, 1),
                         MakeCmdAttr<CommandSInter>("sinter", -2, "read-only", 1, -1, 1),
+                        MakeCmdAttr<CommandSInterCard>("sintercard", -3, "read-only", 1, -1, 1),
```
Comment (infdahai) [inline src/commands/cmd_set.cc]: @PragmaTwice Would it be better to use this change based on the syntax `SINTERCARD numkeys key [key ...] [LIMIT limit]` ? It seems this range is related to `watch`
```suggestion
                        MakeCmdAttr<CommandSInterCard>("sintercard", -3, "read-only", 2, -1, 1),
```

Comment (PragmaTwice) [reply src/commands/cmd_set.cc]: The best solution here is write a `CommandKeyRangeGen` function, since we have to determine the last position of the key sequence. 

And of course, the position of the first key is 2.

Comment (infdahai) [reply src/commands/cmd_set.cc]: Thanks. I've got it.

**Comment by infdahai** [comment]
ready for reviews.

File: src/commands/cmd_set.cc [hunk 9]
```diff
@@ -283,6 +285,65 @@ class CommandSInter : public Commander {
...
+        return {Status::RedisParseErr, errLimitIsNegative};
+      }
+    }
+    return Commander::Parse(args);
+  }
+
+  Status Execute(Server *svr, Connection *conn, std::string *output) override {
+    std::vector<Slice> keys;
+    for (size_t i = 2; i < numkeys_ + 2; i++) {
+      keys.emplace_back(args_[i]);
```
Comment (PragmaTwice) [inline src/commands/cmd_set.cc:339]: It seems there is no check for the boundary of `args_`, e.g. for `SINTERCARD 99 a b` will the `Execute` function just crash on this line since `args_[4]` and so on do not exist?

Comment (infdahai) [reply src/commands/cmd_set.cc:339]: Thank you so much. So many vulnerabilities. But I have a question about this, `go-redis` compute the `num_key` value and pass the all arguments like `redis` commands to the rpc.

Shall we do this check?

Comment (PragmaTwice) [reply src/commands/cmd_set.cc:339]: Any user input via redis protocol cannot just make kvrocks crash. That is the **principal commitment** of development.

And you cannot enforce users to always use go-redis wrapped command functions.

Comment (infdahai) [reply src/commands/cmd_set.cc:339]: OK, I understand this and thanks again.

**Comment by git-hulk** [comment]
The flaky test should be caused by timeout, but didn't check the error in this case: https://github.com/apache/incubator-kvrocks/actions/runs/5070341119/jobs/9105869084#step:12:101

**Comment by infdahai** [comment]
The PR doesn't have complete tests for wrong parsing msgs  based on the latest updating currently. Please wait me for a few days(Sorry, I am moving home).

**Comment by git-hulk** [comment]
@infdahai Thanks, no hurry. Can ping back while your PR is ready to review.

**Comment by infdahai** [comment]
@git-hulk I was resurrected. Ready to review.

File: tests/cppunit/types/set_test.cc [hunk 10]
```diff
@@ -188,9 +188,41 @@ TEST_F(RedisSetTest, Inter) {
...
+  set_->InterCard({k1, k2}, 3, &ret);
+  EXPECT_EQ(ret, 2);
+  set_->InterCard({k2, k3}, 1, &ret);
+  EXPECT_EQ(ret, 1);
+  set_->InterCard({k1, k3}, 5, &ret);
+  EXPECT_EQ(ret, 0);
+  set_->InterCard({k1, k4}, 5, &ret);
+  EXPECT_EQ(ret, 0);
+  set_->InterCard({k1}, 0, &ret);
+  EXPECT_EQ(ret, 4);
```
Comment (torwig) [inline tests/cppunit/types/set_test.cc:227]: If I add the following test:

```
set_->InterCard({k1}, 2, &ret);
EXPECT_EQ(ret, 2);
```

it will fail.

I tested it here: https://redis.io/commands/sintercard/

```
redis> sintercard 1 key1 limit 0
(integer) 4
redis> sintercard 1 key1 limit 2
(integer) 2
```

Comment (infdahai) [reply tests/cppunit/types/set_test.cc:227]: OK, I'll add more tests later in corner cases. The test for a single key has been updated. Thanks a lot.

File: src/types/redis_set.cc [hunk 11]
```diff
@@ -332,29 +332,66 @@ rocksdb::Status Set::Union(const std::vector<Slice> &keys, std::vector<std::stri
...
   members->clear();
 
   std::map<std::string, size_t> member_counters;
   std::vector<std::string> target_members;
   auto s = Members(keys[0], &target_members);
   if (!s.ok() || target_members.empty()) return s;
+
+  uint64_t limit = cnt ? *cnt : 0;
+  bool has_limit = limit != 0;
+  bool limited = false;
```
Comment (torwig) [inline src/types/redis_set.cc]: I think you can rename this variable to `limit_reached`.

**Review by torwig — COMMENTED** [review]
In general, it seems like the functionality is correct.

However, I don't like the implementation because (if I understood properly) the `cnt` parameter of the `Set::Inter` function serves as the sign that this function calculates only cardinality AND passes the limit (which has only the `SINTERCARD` command) AND simultaneously serves as the return value. The idea to introduce a separate function directly is bothering me every time I review this PR.
The disadvantage is obviously the additional function in this case.

I would like to hear @git-hulk , @PragmaTwice on this.

File: src/types/redis_set.cc (hunk 11 above)
Comment (infdahai) [reply src/types/redis_set.cc]: Yeah, I should split the func.

File: src/types/redis_set.cc [hunk 12]
```diff
@@ -332,21 +332,21 @@ rocksdb::Status Set::Union(const std::vector<Slice> &keys, std::vector<std::stri
...
 
-  uint64_t limit = cnt ? *cnt : 0;
   bool has_limit = limit != 0;
-  bool limited = false;
+  bool limit_reached = false;
   size_t keys_size = keys.size();
 
   if (keys_size == 1 && has_limit) {
-    if (limit > target_members.size()) *cnt = target_members.size();
+    *cnt = limit > target_members.size() ? target_members.size() : limit;
```
Comment (PragmaTwice) [inline src/types/redis_set.cc]: Here maybe a nullptr check is needed for `cnt`. But IMHO, we can assume that `cnt` can never be nullptr, remove all these check and the default argument of `cnt` (`nullptr`) in the function declaration.

**Comment by infdahai** [comment]
https://github.com/apache/incubator-kvrocks/actions/runs/5234808336/jobs/9452175246

Do we need to ignore this?

**Comment by PragmaTwice** [comment]
> https://github.com/apache/incubator-kvrocks/actions/runs/5234808336/jobs/9452175246
> 
> Do we need to ignore this?

Nope. The build procedure need to succeed in Darwin. It cannot be ignored.

You can add some type cast before the std::min call to make it compile successfully.

**Comment by torwig** [comment]
@infdahai You didn't actually split the `Inter` function into two separate :) I meant that the `Inter` function should calculate the intersection between sets and return members while `InterCard` should only calculate the cardinality and return the computed value. The command `Inter` will call `Set::Inter` and the command `InterCard` will call the `Set::InterCard` (currently the `Set::InterCard` function calls `Set::Inter` and all the complexity is still in the `Set::Inter` function). In this case, you will get rid of some of the `if` statements and simplify the code.

**Comment by infdahai** [comment]
> @infdahai You didn't actually split the `Inter` function into two separate :) I meant that the `Inter` function should calculate the intersection between sets and return members while `InterCard` should only calculate the cardinality and return the computed value. The command `Inter` will call `Set::Inter` and the command `InterCard` will call the `Set::InterCard` (currently the `Set::InterCard` function calls `Set::Inter` and all the complexity is still in the `Set::Inter` function). In this case, you will get rid of some of the `if` statements and simplify the code.

@torwig I just thought if I split the functions, these two functions exist duplicate codes before and we add the limit `0` to `Set::Inter` to unify the two functions currently. And I also find the generic function is integrated in redis. https://github.com/redis/redis/blob/f228ec1ea5ef8f3734f1dc0297c5181a651aba1d/src/t_set.c#L1276

I remove the if statements but use duplicate `for` statements. Of course, I will do as you say.

File: src/types/redis_set.cc [hunk 13]
```diff
@@ -358,6 +358,54 @@ rocksdb::Status Set::Inter(const std::vector<Slice> &keys, std::vector<std::stri
...
+    s = Members(keys[i], &target_members);
+    if (!s.ok() || target_members.empty()) {
+      return s;
+    }
+
+    for (const auto &member : target_members) {
+      auto iter = member_counters.find(member);
+      if (iter == member_counters.end()) continue;
+      if (++iter->second == keys_size) {
+        cardinality++;
```
Comment (torwig) [inline src/types/redis_set.cc]: Is there any reason to introduce a new variable `cardinality` and later increment it instead of incrementing the `cnt` variable?

Comment (infdahai) [reply src/types/redis_set.cc]: ignore this. I don't like accessing pointers all the time.

**Review by torwig — DISMISSED** [review]
Generally, LGTM.

File: src/types/redis_set.cc [hunk 14]
```diff
@@ -358,6 +358,52 @@ rocksdb::Status Set::Inter(const std::vector<Slice> &keys, std::vector<std::stri
...
+    s = Members(keys[i], &target_members);
+    if (!s.ok() || target_members.empty()) {
+      return s;
+    }
+
+    for (const auto &member : target_members) {
+      auto iter = member_counters.find(member);
+      if (iter == member_counters.end()) continue;
+      if (++iter->second == keys_size) {
+        *cardinality++;
```
Comment (infdahai) [inline src/types/redis_set.cc]:
```suggestion
        *cardinality += 1;
```

**Review by git-hulk — APPROVED** [review]

**Review by torwig — APPROVED** [review]

**Review by PragmaTwice — APPROVED** [review]

**Comment by git-hulk** [comment]
Thanks all, merging...

**Comment by git-hulk** [comment]
@infdahai Can add the SINTERCARD command to https://github.com/apache/incubator-kvrocks-website/blob/main/docs/supported-commands.md
